index.js

```diff
@@ -1,302 +1,293 @@
-    this._rich.push({capabilities, send, media});
-    return this;
-  }
-  promise(func) {
-    let r = func(this);
-    if (!Convo.isPromise(r)) {
-      return Promise.resolve(r);
-    }
-    return r;
-  }
-  setAccessToken(token) {
-    if (!conv.user) {
-      conv.user = {};
-    }
-    if (!conv.user.access) {
-      conv.user.access = {};
-    }
-    conv.user.access.token = token;
-    return this;
-  }
-  getContext(context) {
-    if (this.conv && this.conv.contexts) {
-      var rContext;
-      if (this.conv.contexts.get) {
-        rContext = this.conv.contexts.get(context);
-      } else {
-        rContext = this.conv.contexts[context];
-      }
-      if (rContext && rContext.parameters) {
-        return rContext.parameters;
-      }
-    }
-    return {};
-  }
-  setContext(context, lifespan, parameters) {
-    if (this.conv && this.conv.contexts) {
-      if (this.conv.contexts.set) {
-        this.conv.contexts.set(context, lifespan, parameters);
-      } else {
-        this.conv.contexts[context] = {lifespan, parameters};
-      }
-    }
-    return this;
-  }
-  getStorage() {
-    if (this.conv && this.conv.user && this.conv.user.storage) {
-      return this.conv.user.storage;
-    }
-    return {};
-  }
+	static mockConv() {
+		return {
+			surface: {
+				capabilities: {
+					has: () => true
+				}
+			},
+			available: {
+				surfaces: {
+					capabilities: {
+						has: () => true
+					}
+				}
+			},
+			user: {
+				access: {},
+				storage: {}
+			},
+			contexts: {}
+		};
+	}
+
+	static setClassMappings(mappings) {
+		this._clsMappings = mappings;
+	}
+
+	/**
+	 * @private
+	 */
+	static processMethod(type, data) {
+		return this._clsMappings && this._clsMappings[type] ? this._clsMappings[type](data) : { type, data };
+	}
+
+	/* eslint-disable new-cap */
+
+	static SignIn(obj) {
+		return this.processMethod('SignIn', obj);
+	}
+
+	static SimpleResponse(obj) {
+		return this.processMethod('SimpleResponse', obj);
+	}
+
+	static NewSurface(obj) {
+		return this.processMethod('NewSurface', obj);
+	}
+
+	static Image(obj) {
+		return this.processMethod('Image', obj);
+	}
+
+	static List(obj) {
+		return this.processMethod('List', obj);
+	}
+
+	static BasicCard(obj) {
+		return this.processMethod('BasicCard', obj);
+	}
+
+	static Button(obj) {
+		return this.processMethod('Button', obj);
+	}
+
+	static isNotEmpty(str) {
+		return str && str.trim() !== '';
+	}
+
+	static complete(convo, action = 'close', options = { log: false, logFunc: null }) {
+		if (!convo.conv){
+			if (options.log) {
+				console.log(`${action}:${JSON.stringify(convo, null, 1)}`);
+			}
+			return Promise.resolve(convo);
+		}
+		let finalText = convo._write.reduce((acc, text) => acc.sentence(text), new Say()).toString().trim();
+		let finalSpeech = convo._speak.reduce((acc, text) => acc.sentence(text), new Say()).toString().trim();
+		let promises = [];
+		let textWasPopulated = !!(finalText || finalSpeech);
```

Comment:
In javascript, an empty string is falsey, so we don't need the additional variables to hold `textIsNotEmpty` and `speechIsNotEmpty` as long as you trim the strings above.  If you don't want to trim them for some reason, you can undo the change, but given how you are using them, that doesn't seem necessary

Reply:
That's the Java dev in me not used to "falsey" things =). Thanks for the tip there.